### DISCUSSION TITLE
프로젝트 공격, 체력 같은 처리 관련

### DISCUSSION THREAD

**Discussion by fkdl0048** [fkdl0048/Mentoring-Project#20] [Q&A] [answered]
플레이어 몬스터 보스가 동일한 함수를 가지도록 설계할려면 인터페이스기능을 활용하는게 좋은게 맞을까요?  

지금까지는 프로젝트에서 플레이어 "공격메서드에서 적이 검출된다면 공격처리" 이런식으로 구현을 해놨는데 몬스터 보스도 크게 다른 방식이 아니라면 인터페이스로 매핑을 해놓는게 좋은 방법이라고 알고 있었습니다.  

이번에 구조를 변경해볼려고 하는데 동일한 기능들은 인터페이스로 묶어서 상속받아서 구헌 -> 검출할때도 gameobject.Getcomponet<인터페이스>()이런방식이 더 좋은 방법일까요?

++ 만약 사용하지 않는 함수까지 구현해야한다고 할 때 비워두는건 용납되지 않는 건가요?

**Comment by jongfeel** [comment, marked as answer]
음 항상 그렇게 구현하는게 좋다고도 나쁘다고도 할 수는 없습니다.
인터페이스를 사용하는 건 좋지만, 상속을 자주 사용하다 보면 나중에 고치기가 쉽지는 않거든요.

그래도 인터페이스를 사용하는 경험을 해보지 않으셨다면 한번 구현해 보시는 것도 나쁘지는 않습니다.

사용하지 않는 함수를 구현할 때 비워두는건 그렇지 않게 할 수 있는 문법을 지원해 줍니다.
Unity에서는 될지 모르겠지만요.

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/default-interface-methods

**Reply by fkdl0048** [reply]
객체지향의 개념에서는 상속을 활용하는게 좋다고 알고 있었는데 상속도 너무 자주사용하게 되면 프로젝트를 읽기 힘들어지고 수정이 힘들어진다는 건 알고 있었지만, 안좋다고 하는게 많이 사용하면 안좋은건지 아니면 적절하게 사용하는게 좋은건지, 프로젝트의 성격마다 달라지는건지 조금 어려운 것 같습니다. 

인터페이스기능을 많이 사용해보고 싶어서 한 번 구현해보고 직접 문제를 경험해보고 싶어서 말씀해주신 대로 기능형식의 구조로 가져가보고 문제점을 느껴보겠습니다아..!
